subprojects/core/src/main/java/org/gradle/api/services/internal/BuildServiceProvider.java

```diff
@@ -117,72 +57,39 @@ public Object unpackState() {
+    @SuppressWarnings("unused") // Used via instrumentation
+    public static <P extends BuildServiceParameters, T extends BuildService<P>> void setBuildServiceAsConvention(@Nonnull DefaultProperty<T> property, ServiceLookup serviceLookup, String buildServiceName) {
+        BuildServiceRegistryInternal buildServiceRegistry = (BuildServiceRegistryInternal) serviceLookup.get(BuildServiceRegistry.class);
+        BuildServiceProvider<T, P> consumer = Cast.uncheckedCast(buildServiceRegistry.consume(buildServiceName, property.getType()));
+        property.convention(consumer);
     }
 
-    private T instantiate(ServiceLookup instantiationServices) {
-        return instantiationScheme.withServices(instantiationServices).instantiator().newInstance(implementationType);
-    }
+    public abstract BuildServiceDetails<T, P> getServiceDetails();
 
-    private ServiceLookup instantiationServicesFor(@Nullable P isolatedParameters) {
-        return isolationScheme.servicesForImplementation(
-            isolatedParameters,
-            internalServices,
-            ImmutableList.of(),
-            serviceType -> false
-        );
-    }
+    public abstract String getName();
 
-    @Override
-    public ExecutionTimeValue<? extends T> calculateExecutionTimeValue() {
-        return ExecutionTimeValue.changingValue(this);
-    }
-
-    public void maybeStop() {
-        synchronized (this) {
-            try {
-                if (instance != null) {
-                    instance.ifSuccessful(t -> {
-                        if (t instanceof AutoCloseable) {
-                            try {
-                                ((AutoCloseable) t).close();
-                            } catch (Exception e) {
-                                throw new ServiceLifecycleException("Failed to stop service '" + name + "'.", e);
-                            }
-                        }
-                    });
-                }
-            } finally {
-                instance = null;
-            }
+    /**
+     * Are the given providers referring to the same service provider?
+     *
+     * This method does not distinguish between consumed/registered providers.
+     */
+    public static boolean isSameService(Provider<? extends BuildService<?>> thisProvider, Provider<? extends BuildService<?>> anotherProvider) {
+        if (!(thisProvider instanceof BuildServiceProvider && anotherProvider instanceof BuildServiceProvider)) {
+            return false;
+        }
+        if (thisProvider == anotherProvider) {
+            return true;
         }
+        BuildServiceProvider thisBuildServiceProvider = (BuildServiceProvider) thisProvider;
+        BuildServiceProvider otherBuildServiceProvider = (BuildServiceProvider) anotherProvider;
+        return thisBuildServiceProvider.getName().equals(otherBuildServiceProvider.getName()) && thisBuildServiceProvider.getType() == otherBuildServiceProvider.getType();
```

Comment:
❓ Shouldn't their `buildIdentifier`s be compared here as well?

Reply:
I considered at some point, but could not come up with concrete cases where that could happen (the identifier defines the registry instance a restored service provider is created under). I will add.